### ISSUE TITLE
Feature Request: Generate pipeline using MPTv2 from command line

### ISSUE THREAD

**Issue by linjmeyer** [spinnaker/spin#203]
Creating or updating a pipeline based on a template still requires a pipeline json file which in many cases is just a reference to the managed pipeline and some variables if needed.  At large organizations with many pipelines that is a lot of repositories with nearly identical json files, generally managed by developers who are unfamiliar with the nuances of Spinnaker.

It would be helpful to create/update a templated pipeline using only the `spin` CLI.  This could either generate a new pipeline or simply generate the pipeline json which could be passed to `spin pipeline save`.

Benefits:
* Less json/config files for developers to manage
* Allows CI systems like Jenkins to create pipelines without developer defined configurations

Proposal:
* Add ability to create a pipeline json file for a given template name
* Accept variables and values from command line
* More complex scenarios like overriding a templates triggers or artifacts would not be supported due to the complexity of implementing this in a CLI

Examples:
* `spin pipeline-template use --application myapp --variables var1=value1,var2=value2`
* `spin pipeline generate --template mytemplate --application myapp --variables var1=value1,var2=value2`

**Comment by milesmatthias** [comment]
I'd prefer for this to generate the JSON, rather than the other option of actually creating the pipeline on the spinnaker instance. This would allow users to double check their intended JSON before actually creating the pipelines.

In fact, I'd also love to see another option in your command where you pull down the template JSON and render it as a new pipeline without depending on the template going forward. Something like `--full-hydrate=true` or `--mpt-dependency=false`.

One of the reasons we see customers avoid MPT currently is the limited flexibility after initial creation. If they want to tweak the pipeline after creation when they're generating experimental pipelines, or they're simply early in their spinnaker adoption, they don't want to be locked into a scenario where they have to change the template if they want to change the pipeline.

**Comment by linjmeyer** [comment]
@milesmatthias that's a cool idea, I've run into that myself and ended up copy/pasting the MPT into Spinnaker so I can make a change in the UI, then copy/paste back into the template after.  It gets tedious.

Do you know if a MPT's pipeline definition is always a fully functional pipeline or does Spin/Spinnaker add missing fields in in some cases?   I'm wondering if I need to account for that when doing something like `-fully-hydrate=true` to make sure the generated pipeline isn't missing any required fields.

**Comment by milesmatthias** [comment]
Should always be a fully functional pipeline.

**Comment by louisjimenez** [comment]
Interesting! I would also favor using the command to generate the JSON for further editing and then persisting via `spin pi save`.

@milesmatthias Thanks for sharing that use case. Would using `spin pt plan` to obtain a hyrdated version of the pipeline help here? There would need to be some minor data cleanup of the result such as removing the `templateVariables` object.

Also, is the customer reluctance to change the template coming from the lack of a UI template editor? Do they find it easier just to create new pipelines rather than publish new template versions?

**Comment by milesmatthias** [comment]
Ideally there wouldn't be any data cleanup necessary. The JSON rendered should be something you can pipe directly into an upload.

No, it's lack of being able to change a pipeline that has been created from a template.

**Comment by louisjimenez** [comment]
Agreed, we definitely want to support directly uploading the returned JSON. The changes to the data structure returned from `plan` would be small enough that the functionality could be an option to the existing command.

**Comment by linjmeyer** [comment]
I have a fork with the changes in progress, so far it looks like `spin pt use <tpl_id> --application my-app --name my-pipeline --description my-desc --variables=var1=1,var2=2` Not all those fields are required but you get the idea.

Curious what you guys think of `spin pt use` vs `spin pi plan` or some other entry point?

**Comment by louisjimenez** [comment]
I think it makes sense to keep it grouped with the pipeline-template command since it is directly tied to generating templated pipelines.

@jingsfdc this thread may interest you, covers using spin to generate pipelines.

**Comment by jingsfdc** [comment]
Thanks @louisjimenez for sharing. If I may add a scope or improve the interface to add the ability to accept an input file for the variables. In our use case, we'd like to generate a different pipeline for each target environment we'd like to run the pipeline on. We have a complex set of variables and could use a variable file input rather than setting the variables as part of the cli.

**Comment by linjmeyer** [comment]
Good idea, working on it.  Looking a lot like helm template command so far: `--set mykey=myval --values valuesfile.json,anotherfile.json`

**Comment by jingsfdc** [comment]
@linjmeyer is there a plan to make spin run in a batch mode where it can take multiple pipeline templates and multiple set of var files to generate multiple pipelines? should we file a different feature request for that? or are you including that in the scope for this work, too?

**Comment by linjmeyer** [comment]
@jingsfdc no plans for supporting multiple templates in one command.  Would a loop in a shell script work though?  That would be easily doable with what I have so far
